### DISCUSSION TITLE
Dynamic tests based on response of request.

### DISCUSSION THREAD

**Discussion by ghost** [cypress-io/cypress#16539] [General]
Hello everyone,

I'm trying to find a way where I can generate (smoke)tests based on the response of a request that will give me all the urls to test. There is an example but it is limiting itself to a defined set. That example can be found here: https://github.com/cypress-io/cypress-example-recipes/blob/master/examples/fundamentals__dynamic-tests/cypress/integration/request-spec.js

But in my case I would like to base my foreach off of the actual response data. So my test should be dynamic in the number of the array as well. Right now I'm still limited to generating a fixture before hand and then using that as the list input for the generation of my tests. So what I would love to do is to pass a variable from a before function to the generation of the tests as in:

```js
/// <reference types="cypress" />
describe('Smoketest on list of urls', () => {
  
  before(() => {
    const req = cy.request('http://web/cypress_smoketest/get/urls')
  })
  
  // This is what I'm doing now:
  const urls = require('../fixtures/urls')
  urls.forEach((url) => {
    it(`Visit ${url}`, () => {
      cy.visit(url)
    })
  })

  // This is what I would like to do.
  req.response.data.forEach((url) => {
    it(`Visit ${url}`, () => {
      cy.visit(url)
    })
  })

})
```

Can anyone help me out here? I have a background in PHP, so for me it's not entirely clear how I can pass these variables around through this spec.js file.

**Comment by jennifer-shehane** [comment]
You should be able to follow a similar pattern as demonstrated in the example, just the variable is a url instead of a user. 

Something like

```js
describe('Smoketest on list of urls', () => {
    let urls
    
    before(() => {
        cy.request('http://web/cypress_smoketest/get/urls').then((response) => {
            urls = response.body
        })
        
    })

    Cypress._.range(0, 3).forEach((k) => {
        it(`Visit ${k}`, () => {
            const url = urls[k]

            cy.visit(`${url}`)
        })
    })
})
```

**Reply by ghost** [reply]
Hi @jennifer-shehane ,

Thanks for your response. What I was trying to say is that I'm unable in the example you bring up, to use the urls variable outside of the it() function. Meaning I have no way of dynamically generating a number of tests.

I don't see how I can base the forEach loop on the urls variable. I do not have access to this urls variable outside of this it() function.

I had the same problem in phpunit and behat that it will not accept any dynamic datasets, only fixed datasets. I think many people are not seeing the power of generating tests on the fly I guess.

But I thought I read that Elon musk is doing stuff like that, so I guess it takes genius level skills. Maybe if I drop the list of urls into an environment variable :) All I do is try to find workarounds lol.

**Comment by bahmutov** [comment]
See the solution in https://github.com/cypress-io/cypress-example-recipes/tree/master/examples/fundamentals__dynamic-tests-from-api

- you make the request in the plugin file when the project loads. This gives you a list of URLs
- you save that list in the `config.env.list = ...` property
- the env object will be available in each spec file loaded after that in the browser. The spec can access it _outside_ any test using `Cypress.env('list')` This is the right moment to create individual tests
```js
Cypress.env('list').forEach(url => {
  it(`Visit ${url}`, () => {
      cy.visit(url)
    })
})
```
